In [2]:
!pip install kaggle

You are using pip version 19.0.2, however version 19.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [4]:
#environment reference for the kaggle api key
%env KAGGLE_CONFIG_DIR=/home/mishraka/Documents/Manjula/Lambda_School/Assignments/Unit4_Tree_Ensembles/Kaggle_Homework_Week

env: KAGGLE_CONFIG_DIR=/home/mishraka/Documents/Manjula/Lambda_School/Assignments/Unit4_Tree_Ensembles/Kaggle_Homework_Week


In [6]:
!kaggle competitions download -c ds1-tree-ensembles

 86%|████████████████████████████████▌     | 4.00M/4.67M [00:00<00:00, 7.57MB/s]
100%|██████████████████████████████████████| 4.67M/4.67M [00:00<00:00, 8.48MB/s]
  0%|                                                | 0.00/295k [00:00<?, ?B/s]
100%|████████████████████████████████████████| 295k/295k [00:00<00:00, 10.2MB/s]
 86%|████████████████████████████████▌     | 1.00M/1.17M [00:00<00:00, 8.43MB/s]
100%|██████████████████████████████████████| 1.17M/1.17M [00:00<00:00, 8.47MB/s]
  0%|                                               | 0.00/92.2k [00:00<?, ?B/s]
100%|██████████████████████████████████████| 92.2k/92.2k [00:00<00:00, 12.4MB/s]
  0%|                                               | 0.00/23.0k [00:00<?, ?B/s]
100%|██████████████████████████████████████| 23.0k/23.0k [00:00<00:00, 17.0MB/s]


In [18]:
pwd

'/home/mishraka/Documents/Manjula/Lambda_School/Assignments/Unit4_Tree_Ensembles/Kaggle_Homework_Week'

In [19]:
ls

kaggle.json             sample_submission.csv  train_features.csv
Kaggle_monday_HW.ipynb  test_features.csv      train_features.csv.zip
LCDataDictionary.xlsx   test_features.csv.zip  train_labels.csv


In [44]:
import pandas as pd
pd.set_option('display.max_columns', 105)

In [45]:
df_train_features = pd.read_csv("train_features.csv")

In [46]:
df_train_features.head()

id  member_id  loan_amnt  funded_amnt        term int_rate  installment  \
0  43373        NaN      12000        12000   36 months   16.02%       422.01   
1  24414        NaN       6300         6300   36 months   14.07%       215.54   
2  46723        NaN       4500         4500   36 months    7.21%       139.38   
3  24878        NaN      12000        12000   36 months    9.44%       384.06   
4  39038        NaN      12000        12000   36 months   22.35%       460.47   

  grade sub_grade                    emp_title emp_length home_ownership  \
0     C        C5                       Driver    4 years       MORTGAGE   
1     C        C3                     GMP Lead    4 years           RENT   
2     A        A3                    Accounts     3 years           RENT   
3     B        B1  Client Processing Associate  10+ years       MORTGAGE   
4     D        D5       Construction Operator   10+ years       MORTGAGE   

   annual_inc  url  desc             purpose               title zip_code  \
0     81000.0  NaN   NaN  debt_consolidation  Debt consolidation    280xx   
1     39000.0  NaN   NaN  debt_consolidation  Debt consolidation    920xx   
2     78000.0  NaN   NaN  debt_consolidation  Debt consolidation    906xx   
3     62000.0  NaN   NaN  debt_consolidation  Debt consolidation    441xx   
4     70000.0  NaN   NaN  debt_consolidation  Debt consolidation    751xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NC  12.76            0         Dec-2005               1   
1         CA  21.42            0         Mar-2012               1   
2         CA   2.17            0         May-2000               2   
3         OH  13.76            0         Feb-2003               0   
4         TX  14.04            4         Nov-2001               2   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                    48.0                     NaN         7        0   
1                     NaN                     NaN        10        0   
2                     NaN                     NaN        13        0   
3                     NaN                     NaN        22        0   
4                    13.0                     NaN        18        0   

   revol_bal revol_util  total_acc initial_list_status  \
0       4912      23.5%         15                   w   
1      11876      59.4%         12                   w   
2       1715       5.2%         19                   w   
3       9404      16.3%         33                   w   
4      17895      24.2%         45                   w   

   collections_12_mths_ex_med  mths_since_last_major_derog application_type  \
0                           0                         48.0       Individual   
1                           0                          NaN       Individual   
2                           0                          NaN       Individual   
3                           0                          NaN       Individual   
4                           0                         13.0       Individual   

   annual_inc_joint  dti_joint  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0               NaN        NaN               0             0       174783   
1               NaN        NaN               0             0        31046   
2               NaN        NaN               0             0        35329   
3               NaN        NaN               0             0        75160   
4               NaN        NaN               0          7218       113702   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0            1            2            0            4                13.0   
1            0            1            0            2                14.0   
2            0            8            1            2                10.0   
3            4            1            0            2                13.0   
4            9            2            2            4                 6.0   

 

In [24]:
df_test_features = pd.read_csv("test_features.csv")

In [25]:
df_test_features.head()

id  member_id  loan_amnt  funded_amnt        term int_rate  installment  \
0  57146        NaN      18500        18500   60 months   25.82%       551.93   
1  20988        NaN      36000        36000   60 months   12.61%       811.95   
2  10031        NaN      12000        12000   36 months   22.90%       463.90   
3  33892        NaN      24875        24875   60 months   11.98%       553.08   
4  35374        NaN       4000         4000   36 months   10.91%       130.79   

  grade sub_grade                       emp_title         ...           \
0     E        E4  Account Service Representative         ...            
1     C        C1            Vice President of IT         ...            
2     E        E1          Director of Technology         ...            
3     B        B5                              vp         ...            
4     B        B4               Sales consultant          ...            

  sec_app_inq_last_6mths sec_app_mort_acc  sec_app_open_acc  \
0                    NaN              NaN               NaN   
1                    NaN              NaN               NaN   
2                    NaN              NaN               NaN   
3                    NaN              NaN               NaN   
4                    0.0              2.0               7.0   

   sec_app_revol_util  sec_app_open_act_il sec_app_num_rev_accts  \
0                 NaN                  NaN                   NaN   
1                 NaN                  NaN                   NaN   
2                 NaN                  NaN                   NaN   
3                 NaN                  NaN                   NaN   
4                22.9                  3.0                   3.0   

  sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med  \
0                              NaN                                NaN   
1                              NaN                                NaN   
2                              NaN                                NaN   
3                              NaN                                NaN   
4                              0.0                                0.0   

  sec_app_mths_since_last_major_derog  disbursement_method  
0                                 NaN                 Cash  
1                                 NaN                 Cash  
2                                 NaN                 Cash  
3                                 NaN                 Cash  
4                                 NaN                 Cash  

[5 rows x 103 columns]

In [26]:
df_test_features.shape, df_train_features.shape

((9437, 103), (37745, 103))

In [27]:
df_train_labels = pd.read_csv("train_labels.csv")

In [47]:
df_train_labels.shape

(37745, 2)

In [111]:
df_test_features.shape

(9437, 103)

In [28]:
df_train_labels.head()

id  charged_off
0  43373            0
1  24414            0
2  46723            0
3  24878            0
4  39038            0

In [30]:
df_sample_submission = pd.read_csv("sample_submission.csv")

In [31]:
df_sample_submission.head()

id  charged_off
0  57146          0.5
1  20988          0.5
2  10031          0.5
3  33892          0.5
4  35374          0.5

In [32]:
df_sample_submission.shape

(9437, 2)

### The total rows in our training set are 37745 

In [113]:
df_train_test_features = pd.concat([df_train_features, df_test_features])

In [115]:
df_train_test_features.shape

(47182, 103)

In [114]:
df_train_test_features.isna().sum()

id                                         0
member_id                              47182
loan_amnt                                  0
funded_amnt                                0
term                                       0
int_rate                                   0
installment                                0
grade                                      0
sub_grade                                  0
emp_title                               4411
emp_length                              4062
home_ownership                             0
annual_inc                                 0
url                                    47182
desc                                   47182
purpose                                    0
title                                      0
zip_code                                   0
addr_state                                 0
dti                                      112
delinq_2yrs                                0
earliest_cr_line                           0
inq_last_6

In [33]:
!pip install dtreeviz

  Using cached https://files.pythonhosted.org/packages/1f/e2/ef2581b5b86625657afd32030f90cf2717456c1d2b711ba074bf007c0f1a/graphviz-0.10.1-py2.py3-none-any.whl
  Stored in directory: /home/mishraka/.cache/pip/wheels/4c/3f/87/b2c828617d1bd181145fbaf41129e4c82fa8d9a98082701e00
Successfully built dtreeviz
You are using pip version 19.0.2, however version 19.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [34]:
!pip install --upgrade pip

  Using cached https://files.pythonhosted.org/packages/d8/f3/413bab4ff08e1fc4828dfc59996d721917df8e8583ea85385d51125dceff/pip-19.0.3-py2.py3-none-any.whl
  Found existing installation: pip 19.0.2
    Uninstalling pip-19.0.2:
      Successfully uninstalled pip-19.0.2


In [36]:
!pip install category_encoders

In [116]:
df_train_test_features.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'emp_title',
       ...
       'sec_app_inq_last_6mths', 'sec_app_mort_acc', 'sec_app_open_acc',
       'sec_app_revol_util', 'sec_app_open_act_il', 'sec_app_num_rev_accts',
       'sec_app_chargeoff_within_12_mths',
       'sec_app_collections_12_mths_ex_med',
       'sec_app_mths_since_last_major_derog', 'disbursement_method'],
      dtype='object', length=103)

### Let's drop the columns with NaNs from the merged file

In [117]:
df_no_na = df_train_test_features.dropna(axis=1, how='any')

In [118]:
df_no_na.shape

(47182, 66)

In [119]:
df_no_na.isna().sum()

id                            0
loan_amnt                     0
funded_amnt                   0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
home_ownership                0
annual_inc                    0
purpose                       0
title                         0
zip_code                      0
addr_state                    0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
open_acc                      0
pub_rec                       0
revol_bal                     0
total_acc                     0
initial_list_status           0
collections_12_mths_ex_med    0
application_type              0
acc_now_delinq                0
tot_coll_amt                  0
tot_cur_bal                   0
open_acc_6m                   0
open_act_il                   0
open_il_12m                   0
                             ..
inq_fi  

In [120]:
df_no_na.dtypes.unique()

array([dtype('int64'), dtype('O'), dtype('float64')], dtype=object)

### let select only numeric columns

In [121]:
import numpy as np
df_train_test_all_numeric = df_no_na.select_dtypes(np.number)

In [122]:
df_train_test_all_numeric.shape

(47182, 53)

In [123]:
df_train_test_all_numeric.head()

id  loan_amnt  funded_amnt  installment  annual_inc  delinq_2yrs  \
0  43373      12000        12000       422.01     81000.0            0   
1  24414       6300         6300       215.54     39000.0            0   
2  46723       4500         4500       139.38     78000.0            0   
3  24878      12000        12000       384.06     62000.0            0   
4  39038      12000        12000       460.47     70000.0            4   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0               1         7        0       4912         15   
1               1        10        0      11876         12   
2               2        13        0       1715         19   
3               0        22        0       9404         33   
4               2        18        0      17895         45   

   collections_12_mths_ex_med  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0                           0               0             0       174783   
1                           0               0             0        31046   
2                           0               0             0        35329   
3                           0               0             0        75160   
4                           0               0          7218       113702   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  total_bal_il  \
0            1            2            0            4         33668   
1            0            1            0            2         19170   
2            0            8            1            2         33614   
3            4            1            0            2         17123   
4            9            2            2            4         16711   

   open_rv_12m  open_rv_24m  max_bal_bc  total_rev_hi_lim  inq_fi  \
0            2            5        2962             20900       1   
1            3            8        5153             20000       1   
2            1            1        1196             33300       4   
3            6            9        6898             57600       3   
4            9           10        5373             73900       3   

   total_cu_tl  inq_last_12m  acc_open_past_24mths  chargeoff_within_12_mths  \
0            1             2                     9                         0   
1            0             1                    10                         0   
2            0             3                     3                         0   
3            1             2                    12                         0   
4            6             3                    14                         0   

   delinq_amnt  mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  \
0            0                    23                      1               1   
1            0                    72                      8               8   
2            0                   218                     10              10   
3            0                   179                      1               1   
4            0                   201                      0               0   

   mort_acc  num_accts_ever_120_pd  num_actv_bc_tl  num_actv_rev_tl  \
0         1                      2               2                2   
1         0                      0               6                7   
2         0                      0               3                3   
3         1                      0               1                3   
4         1                      6               9               11   

   num_bc_sats  num_bc_tl  num_il_tl  num_op_rev_tl  num_rev_accts  \
0            2          2          9              4              5   
1            7          8          2              9             10   
2            3          4         13              5              6   
3            9         13          7             20             25   
4           11         14         22             15             20   

   num_rev_tl_bal_gt_0  num_sats  num_tl_30dpd  num_tl_90g_dpd_24m  \
0              

In [96]:
# df_y = df_merged_all_numeric.iloc[:,0:2]

In [124]:
# df_y.head()

In [94]:
# X = df_merged_all_numeric.iloc[:, 2:56]

### Baseline Model 
Let's split our data back into training and test

In [129]:
X_cleaned = df_train_test_all_numeric[:-9437]
X_test_original_shape = df_train_test_all_numeric[-9437:]
y = df_train_labels['charged_off']

In [130]:
X_cleaned.shape, X_test_original_shape.shape, y.shape

((37745, 53), (9437, 53), (37745,))

### train a decision tree

In [144]:
from sklearn.tree import DecisionTreeClassifier 

dt = DecisionTreeClassifier()
dt.fit(X_cleaned,y)
dt.score(X_cleaned,y)

1.0

In [132]:
y_pred = dt.predict_proba(X_test_original_shape)

In [136]:
y_pred.shape

(9437, 2)

In [139]:
sample_submission = pd.read_csv('sample_submission.csv') 
first_submission= sample_submission.copy()
first_submission.charged_off = y_pred
first_submission.to_csv('manjula_first.csv', index=False)

In [147]:
# from dtreeviz.trees import dtreeviz
# dtreeviz(dt, X_cleaned, y,
#         target_name='loan paid',
#         feature_names=X_cleaned.columns,
#         class_names=['No', 'Yes'])

### Working with non numeric columns

In [148]:
df_train_features.shape

(37745, 103)

In [210]:
df_no_na = df_train_test_features.dropna(axis=1, how='any')

In [150]:
df_no_na.shape

(47182, 66)

In [152]:
df_no_na.isna().sum()

id                            0
loan_amnt                     0
funded_amnt                   0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
home_ownership                0
annual_inc                    0
purpose                       0
title                         0
zip_code                      0
addr_state                    0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
open_acc                      0
pub_rec                       0
revol_bal                     0
total_acc                     0
initial_list_status           0
collections_12_mths_ex_med    0
application_type              0
acc_now_delinq                0
tot_coll_amt                  0
tot_cur_bal                   0
open_acc_6m                   0
open_act_il                   0
open_il_12m                   0
                             ..
inq_fi  

In [154]:
df_no_na.dtypes

id                              int64
loan_amnt                       int64
funded_amnt                     int64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
home_ownership                 object
annual_inc                    float64
purpose                        object
title                          object
zip_code                       object
addr_state                     object
delinq_2yrs                     int64
earliest_cr_line               object
inq_last_6mths                  int64
open_acc                        int64
pub_rec                         int64
revol_bal                       int64
total_acc                       int64
initial_list_status            object
collections_12_mths_ex_med      int64
application_type               object
acc_now_delinq                  int64
tot_coll_amt                    int64
tot_cur_bal 

In [155]:
df_no_na.head()

id  loan_amnt  funded_amnt        term int_rate  installment grade  \
0  43373      12000        12000   36 months   16.02%       422.01     C   
1  24414       6300         6300   36 months   14.07%       215.54     C   
2  46723       4500         4500   36 months    7.21%       139.38     A   
3  24878      12000        12000   36 months    9.44%       384.06     B   
4  39038      12000        12000   36 months   22.35%       460.47     D   

  sub_grade home_ownership  annual_inc             purpose  \
0        C5       MORTGAGE     81000.0  debt_consolidation   
1        C3           RENT     39000.0  debt_consolidation   
2        A3           RENT     78000.0  debt_consolidation   
3        B1       MORTGAGE     62000.0  debt_consolidation   
4        D5       MORTGAGE     70000.0  debt_consolidation   

                title zip_code addr_state  delinq_2yrs earliest_cr_line  \
0  Debt consolidation    280xx         NC            0         Dec-2005   
1  Debt consolidation    920xx         CA            0         Mar-2012   
2  Debt consolidation    906xx         CA            0         May-2000   
3  Debt consolidation    441xx         OH            0         Feb-2003   
4  Debt consolidation    751xx         TX            4         Nov-2001   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0               1         7        0       4912         15   
1               1        10        0      11876         12   
2               2        13        0       1715         19   
3               0        22        0       9404         33   
4               2        18        0      17895         45   

  initial_list_status  collections_12_mths_ex_med application_type  \
0                   w                           0       Individual   
1                   w                           0       Individual   
2                   w                           0       Individual   
3                   w                           0       Individual   
4                   w                           0       Individual   

   acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_act_il  \
0               0             0       174783            1            2   
1               0             0        31046            0            1   
2               0             0        35329            0            8   
3               0             0        75160            4            1   
4               0          7218       113702            9            2   

   open_il_12m  open_il_24m  total_bal_il  open_rv_12m  open_rv_24m  \
0            0            4         33668            2            5   
1            0            2         19170            3            8   
2            1            2         33614            1            1   
3            0            2         17123            6            9   
4            2            4         16711            9           10   

   max_bal_bc  total_rev_hi_lim  inq_fi  total_cu_tl  inq_last_12m  \
0        2962             20900       1            1             2   
1        5153             20000       1            0             1   
2        1196             33300       4            0             3   
3        6898             57600       3            1             2   
4        5373             73900       3            6             3   

   acc_open_past_24mths  chargeoff_within_12_mths  delinq_amnt  \
0                     9                         0            0   
1                    10                         0            0   
2                     3                         0            0   
3                    12                         0            0   
4                    14                         0            0   

   mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  \
0                    23                      1               1         1   
1                    72                      8               8         0   
2            

In [157]:
df_no_na.int_rate.value_counts()

 11.98%    1003
 10.90%     925
 12.61%     909
  9.43%     903
 18.45%     888
  9.92%     880
 17.47%     857
  7.34%     836
 13.58%     829
 14.07%     825
 15.04%     819
 21.85%     780
 11.99%     768
 19.42%     758
  5.31%     755
  9.44%     743
 20.39%     741
  6.71%     733
 16.01%     728
 15.05%     723
 12.62%     700
 13.59%     700
  7.96%     694
  9.93%     692
 10.42%     692
 14.08%     686
  7.35%     686
 10.41%     682
  6.07%     661
 16.02%     650
           ... 
 23.40%     167
 22.90%     167
 24.85%     162
 23.87%     146
 26.30%     142
 25.82%     133
 26.31%     132
 24.37%     121
 25.34%     120
  8.19%     111
 11.80%     107
 10.72%      96
 30.79%      95
 12.98%      92
 22.91%      84
  8.81%      83
 10.33%      82
 30.17%      76
 29.69%      73
  7.56%      72
 11.31%      65
 30.75%      60
 30.65%      59
  6.46%      58
  7.02%      56
 30.89%      10
 30.84%       8
 30.94%       5
  6.00%       4
 30.99%       1
Name: int_rate, Length: 

### Getting rid of the % sign from int_rate column

In [172]:
df_no_na['int_rate_float'] = df_no_na.int_rate.str.replace('%','')

/home/mishraka/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


### Converting the new int_rate_float column to float data type

In [170]:
df_no_na['int_rate_float'] = df_no_na['int_rate_float'].astype(float)

/home/mishraka/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [171]:
df_no_na.dtypes

id                              int64
loan_amnt                       int64
funded_amnt                     int64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
home_ownership                 object
annual_inc                    float64
purpose                        object
title                          object
zip_code                       object
addr_state                     object
delinq_2yrs                     int64
earliest_cr_line               object
inq_last_6mths                  int64
open_acc                        int64
pub_rec                         int64
revol_bal                       int64
total_acc                       int64
initial_list_status            object
collections_12_mths_ex_med      int64
application_type               object
acc_now_delinq                  int64
tot_coll_amt                    int64
tot_cur_bal 

In [173]:
df_no_na=df_no_na.drop(columns='int_rate')

In [176]:
df_no_na.shape

(47182, 66)

### Fitting a second model

In [192]:
from sklearn.preprocessing import LabelEncoder
def dummyEncode(df):
        columnsToEncode = list(df.select_dtypes(include=['category','object']))
        le = LabelEncoder()
        for feature in columnsToEncode:
            try:
                df[feature] = le.fit_transform(df[feature])
            except:
                print('Error encoding '+feature)
        return df

In [204]:
dummyEncode(df_no_na)

id  loan_amnt  funded_amnt  term  installment  grade  sub_grade  \
0     43373      12000        12000     0       422.01      2         14   
1     24414       6300         6300     0       215.54      2         12   
2     46723       4500         4500     0       139.38      0          2   
3     24878      12000        12000     0       384.06      1          5   
4     39038      12000        12000     0       460.47      3         19   
5     36809      35000        35000     0      1221.71      2         13   
6     24005      18000        18000     1       407.08      1          9   
7     13722       3600         3600     0       119.56      1          9   
8     26082      25600        25600     1       590.11      2         11   
9     16701      15600        15600     0       479.69      0          2   
10    11504      35000        35000     0      1053.86      0          0   
11    49127      40000        40000     0      1292.01      1          6   
12    14626       6000         6000     0       231.95      4         20   
13    23189      13000        13000     1       276.73      1          5   
14    19667      40000        40000     0      1229.79      0          2   
15    39343      18000        18000     1       406.07      2         10   
16    27322       5000         5000     0       178.49      3         15   
17    34549      10000        10000     1       233.10      2         12   
18    21211      13500        13500     0       419.69      0          3   
19    38196      10000        10000     0       335.07      2         10   
20    36648       7200         7200     0       232.60      1          5   
21    24423       4000         4000     0       128.01      1          5   
22    37588      17000        17000     0       661.62      4         20   
23    41744      15000        15000     0       480.08      1          5   
24    39727       8500         8500     0       278.53      1          7   
25    42693       3000         3000     0        98.09      1          8   
26    23510       8000         8000     0       264.00      1          8   
27    41061      20000        20000     1       451.19      2         10   
28    54681       7000         7000     0       225.65      1          6   
29    31437      35000        35000     0      1146.85      1          7   
...     ...        ...          ...   ...          ...    ...        ...   
9407  11564      12500        12500     1       379.98      4         24   
9408  52626       4000         4000     0       137.63      2         11   
9409  52476      10000        10000     0       323.05      1          5   
9410  17076       6500         6500     0       251.28      4         20   
9411  12684       5000         5000     0       170.97      2         11   
9412  51035       2500         2500     0        76.87      0          2   
9413  15603      34000        34000     1       791.65      2         11   
9414  55140      10000        10000     0       341.93      2         11   
9415  56009      21100        21100     0       842.74      4         22   
9416  40710      16000        16000     0       515.68      1          6   
9417  43809      15000        15000     0       557.46      3         18   
9418  49345       3000         3000     0        96.92      1          5   
9419  10024       8000         8000     0       261.54      1          8   
9420  33850      20000        20000     1       529.88      3         18   
9421  31945       6000         6000     0       215.33      3         15   
9422  26552      20000        20000     0       737.38      3         17   
9423  27546      21325        21325     1       469.53      1          8   
9424  14835      15000        15000     1       321.75      1          7   
9425  26876      35000        35000     1       889.92      3         16   
9426  27344       7200         7200     0       253.63      2         13   
9427  45476      15000        15000     0       490.45   

In [195]:
df_no_na.dtypes

id                              int64
loan_amnt                       int64
funded_amnt                     int64
term                            int64
installment                   float64
grade                           int64
sub_grade                       int64
home_ownership                  int64
annual_inc                    float64
purpose                         int64
title                           int64
zip_code                        int64
addr_state                      int64
delinq_2yrs                     int64
earliest_cr_line                int64
inq_last_6mths                  int64
open_acc                        int64
pub_rec                         int64
revol_bal                       int64
total_acc                       int64
initial_list_status             int64
collections_12_mths_ex_med      int64
application_type                int64
acc_now_delinq                  int64
tot_coll_amt                    int64
tot_cur_bal                     int64
open_acc_6m 

In [203]:
X = df_no_na[:-9437]
X_test = df_no_na[-9437:]
y = df_train_labels['charged_off']

dt = DecisionTreeClassifier(max_depth=3)
dt.fit(X,y)
dt.score(X,y)

0.8528546827394357

In [202]:
df_no_na.head()

id  loan_amnt  funded_amnt  term  installment  grade  sub_grade  \
0  43373      12000        12000     0       422.01      2         14   
1  24414       6300         6300     0       215.54      2         12   
2  46723       4500         4500     0       139.38      0          2   
3  24878      12000        12000     0       384.06      1          5   
4  39038      12000        12000     0       460.47      3         19   

   home_ownership  annual_inc  purpose  title  zip_code  addr_state  \
0               1     81000.0        2      3       253          26   
1               3     39000.0        2      3       792           4   
2               3     78000.0        2      3       779           4   
3               1     62000.0        2      3       403          34   
4               1     70000.0        2      3       652          42   

   delinq_2yrs  earliest_cr_line  inq_last_6mths  open_acc  pub_rec  \
0            0               141               1         7        0   
1            0               406               1        10        0   
2            0               444               2        13        0   
3            0               189               0        22        0   
4            4               495               2        18        0   

   revol_bal  total_acc  initial_list_status  collections_12_mths_ex_med  \
0       4912         15                    1                           0   
1      11876         12                    1                           0   
2       1715         19                    1                           0   
3       9404         33                    1                           0   
4      17895         45                    1                           0   

   application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  \
0                 0               0             0       174783            1   
1                 0               0             0        31046            0   
2                 0               0             0        35329            0   
3                 0               0             0        75160            4   
4                 0               0          7218       113702            9   

   open_act_il  open_il_12m  open_il_24m  total_bal_il  open_rv_12m  \
0            2            0            4         33668            2   
1            1            0            2         19170            3   
2            8            1            2         33614            1   
3            1            0            2         17123            6   
4            2            2            4         16711            9   

   open_rv_24m  max_bal_bc  total_rev_hi_lim  inq_fi  total_cu_tl  \
0            5        2962             20900       1            1   
1            8        5153             20000       1            0   
2            1        1196             33300       4            0   
3            9        6898             57600       3            1   
4           10        5373             73900       3            6   

   inq_last_12m  acc_open_past_24mths  chargeoff_within_12_mths  delinq_amnt  \
0             2                     9                         0            0   
1             1                    10                         0            0   
2             3                     3                         0            0   
3             2                    12                         0            0   
4             3                    14                         0            0   

   mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  \
0                    23                      1               1         1   
1                    72                      8               8         0   
2                   218                     10              10         0   
3                   179                      1               1         1   
4                   201                      0               0         1   

 

In [209]:
df_no_na= df_no_na.drop(columns=['id'])

In [200]:
y_pred = dt.predict_proba(X_test)

In [198]:
y_pred.shape

(9437, 2)

In [207]:
sample_submission = pd.read_csv('sample_submission.csv') 
submission = sample_submission.copy() 
submission['charged_off'] =dt.predict_proba(X_test)[:, 1] 
submission.to_csv('submission-001.csv', index=False)

In [187]:
import category_encoders as ce
import pandas as pd
from sklearn.pipeline import make_pipeline
from sklearn.tree import DecisionTreeClassifier

X = df_no_na[:-9437]
X_test = df_no_na[-9437:]
y = df_train_labels['charged_off']

pipe = make_pipeline(
    ce.OneHotEncoder(use_cat_names=True),
    DecisionTreeClassifier(max_depth=2)
)

In [188]:
pipe.fit(X, y)

Pipeline(memory=None,
     steps=[('onehotencoder', OneHotEncoder(cols=['term', 'grade', 'sub_grade', 'home_ownership', 'purpose', 'title', 'zip_code', 'addr_state', 'earliest_cr_line', 'initial_list_status', 'application_type', 'disbursement_method', 'int_rate_float'],
       drop_invariant=False, handle_unknown='impute', im...      min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'))])

In [189]:
pipe.score(X,y)

0.8528546827394357

In [190]:
y_pred = pipe.predict_proba(X_test)

In [191]:
# sample_submission = pd.read_csv('sample_submission.csv') 
first_submission= sample_submission.copy()
first_submission.charged_off = y_pred
first_submission.to_csv('manjula_forth.csv', index=False)